File: cpp/src/Order.cpp
#include "Order.h"
#include <iostream>
#include <iomanip>
#include <cstdlib>//генерация случайных чисел
#include <ctime>
#include <fstream>
#include <sstream>
#include <limits>//пределы типов данных

using namespace std;

#ifdef _MSC_VER
#define _CRT_SECURE_NO_WARNINGS
#endif

//конструктор по умолчанию
Order::Order() : itemCount(0), countOrder(0), costOrder(0.0) {//создание "пустого" заказа
    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }
    numberOrder = generateOrderNumber();
    statusOrder = "Ожидает оплаты";
    dateOrder = getCurrentDate();
    shopOrder = "Музыкальный магазин 'MusicLoverzz'";
    payWay = "Не выбран";
}

// Конструктор с параметрами (клиент и корзина)
Order::Order(const std::string& clientId, Basket& basket) :
    clientOrder(clientId), itemCount(0), countOrder(0), costOrder(0.0) {

    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }

    numberOrder = generateOrderNumber();
    statusOrder = "Ожидает оплаты";
    dateOrder = getCurrentDate();
    shopOrder = "Музыкальный магазин 'MusicLoverzz'";
    payWay = "Не выбран";

    CopyProductsFromBasket(basket);//копирование информации из корзины
}

// Деструктор класса Order - освобождение памяти
Order::~Order() {
    // Проходим по всем товарам в заказе
    for (int i = 0; i < itemCount; i++) {
        // Проверяем что указатель не нулевой
        if (orderItems[i] != nullptr) {
            delete orderItems[i];// освобождаем память каждого товара
        }
    }
}

// Метод копирования товаров из корзины в заказ
void Order::CopyProductsFromBasket(Basket& basket) {
    if (basket.IsEmpty()) {//проверка не пустая ли корзина
        return;// если корзина пуста - выходим
    }

    // Копируем основную информацию из корзины
    itemCount = basket.getProductCount();// количество различных товаров
    countOrder = basket.GetBasketCount();// общее количество единиц товара
    costOrder = basket.GetBasketCost();// общая стоимость

    // Копируем каждый товар из корзины в заказ
    for (int i = 0; i < itemCount; i++) {
        // Получаем товар из корзины
        Product* basketProduct = basket.getProduct(i);
        // Проверяем что товар существует
        if (basketProduct != nullptr) {
            // Создаем копию товара
            orderItems[i] = new Product(
                // Создаем ПОЛНУЮ КОПИЮ товара с помощью конструктора
                basketProduct->getArticle(),
                basketProduct->getName(),
                basketProduct->getBrand(),
                basketProduct->getProducerCountry(),
                basketProduct->getCategory(),
                basketProduct->getTypeProduct(),
                basketProduct->getDescription(),
                basketProduct->getQuantityStock(),
                basketProduct->getPrice()
            );
            // Копируем количество товара
            quantities[i] = basket.getQuantity(i);
        }
    }
}
// Статический метод для создания заказа (можно вызывать без объекта)
bool Order::MakeOrder(const std::string& clientId, Basket& basket) {
    // Проверка что корзина не пустая
    if (basket.IsEmpty()) {
        cout << "Корзина пуста! Добавьте товары перед оформлением заказа." << endl;
        return false;// возвращаем false если корзина пуста
    }

    // Запрос подтверждения у пользователя
    char confirmation;
    cout << "\n=== ПОДТВЕРЖДЕНИЕ ЗАКАЗА ===" << endl;
    basket.DisplayBasket();
    cout << "\nВы точно хотите оформить заказ? (y/n): ";
    cin >> confirmation;

    if (confirmation != 'y' && confirmation != 'Y') {
        cout << "Оформление заказа отменено." << endl;
        return false;
    }

    // Создаем новый заказ из корзины
    Order* newOrder = new Order(clientId, basket);

    // Проверяем оплату
    if (newOrder->Payment()) {
        // Если оплата прошла успешно
        newOrder->ChangeStatus("Оплачен");
        cout << "\n=== ЗАКАЗ УСПЕШНО ОФОРМЛЕН ===" << endl;
        newOrder->InfoOrder();// показываем информацию о заказе
        newOrder->SaveToFile();// сохраняем заказ в файл

        // Очищаем корзину
        basket.ClearBasket();
        delete newOrder; // освобождаем память
        return true;
    }
    else {
        newOrder->ChangeStatus("Ошибка оплаты");
        cout << "Ошибка при оформлении заказа!" << endl;
        delete newOrder; // освобождаем память
        return false;
    }
}

// Метод сохранения заказа в файл
void Order::SaveToFile() {
    // Открываем файл для добавления
    ofstream file("orders.txt", ios::app);
    if (file.is_open()) {
        // Записываем основную информацию о заказе
        file << numberOrder << "|" << clientOrder << "|" << statusOrder << "|" << costOrder;
        // Записываем информацию о каждом товаре в заказе
        for (int i = 0; i < itemCount; i++) {
            if (orderItems[i] != nullptr) {
                file << "|" << orderItems[i]->getArticle() << "|" << quantities[i];//записать артикул, количество
            }
        }
        file << endl;
        file.close();
    }
}

// Метод изменения статуса заказа
void Order::ChangeStatus(const std::string& newStatus) {
    statusOrder = newStatus;
    if (newStatus == "Оплачен") {
        cout << "\n=== ВАШ ЗАКАЗ УСПЕШНО ОФОРМЛЕН! ===" << endl;
        cout << "Номер заказа: " << numberOrder << endl;
        cout << "Статус: " << statusOrder << endl;
        cout << "==================================" << endl;
    }
    else {
        cout << "Статус заказа изменен на: " << statusOrder << endl; // Для других статусов просто выводим сообщение
    }
}

// Метод обработки оплаты заказа
bool Order::Payment() {
    cout << "\n=== ОПЛАТА ЗАКАЗА ===" << endl;
    cout << "Сумма к оплате: " << costOrder << " руб." << endl;

    int paymentChoice;// хранит число, которое введет пользователь
    bool validChoice = false;// изначально false - выбор еще не сделан или неверен

    while (!validChoice) {
        cout << "Выберите способ оплаты:" << endl;
        cout << "1. Банковская карта" << endl;
        cout << "2. Электронный кошелек" << endl;
        cout << "3. Наложенный платеж" << endl;
        cout << "Ваш выбор: ";
        cin >> paymentChoice;

        if (cin.fail()) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Ошибка! Введите число от 1 до 3." << endl;
            continue;
        }
        // Проверка допустимости выбора
        if (paymentChoice >= 1 && paymentChoice <= 3) {
            validChoice = true;// выбор корректен
        }
        else {
            cout << "Неверный выбор! Введите 1, 2 или 3." << endl;
        }
    }
    // Установка способа оплаты в зависимости от выбора
    switch (paymentChoice) {
    case 1:
        payWay = "Банковская карта";
        break;
    case 2:
        payWay = "Электронный кошелек";
        break;
    case 3:
        payWay = "Наложенный платеж";
        break;
    }

    cout << "Обработка платежа..." << endl;
    cout << "Способ оплаты: " << payWay << endl;
    cout << "Оплата прошла успешно!" << endl;
    return true;
}
// Метод вывода полной информации о заказе
void Order::InfoOrder() const {
    cout << "\n=== ИНФОРМАЦИЯ О ЗАКАЗЕ ===" << endl;
    cout << "Номер заказа: " << numberOrder << endl;
    cout << "Клиент: " << clientOrder << endl;
    cout << "Дата заказа: " << dateOrder << endl;
    cout << "Статус: " << statusOrder << endl;
    cout << "Способ оплаты: " << payWay << endl;
    cout << "Количество товаров: " << countOrder << endl;
    cout << "Общая стоимость: " << costOrder << " руб." << endl;
    cout << "Магазин: " << shopOrder << endl;
    // Вывод состава заказа если есть товары
    if (itemCount > 0) {
        cout << "\nСостав заказа:" << endl;
        cout << "----------------------------------" << endl;
        // Перебор всех товаров в заказе
        for (int i = 0; i < itemCount; i++) {
            if (orderItems[i] != nullptr) {
                // Вывод информации о каждом товаре
                cout << (i + 1) << ". " << orderItems[i]->getName()
                    << " - " << quantities[i] << " шт. x "
                    << orderItems[i]->getPrice() << " руб. = "
                    << (quantities[i] * orderItems[i]->getPrice()) << " руб." << endl;
            }
        }
        cout << "----------------------------------" << endl;
    }
    cout << "==================================" << endl;
}

// Метод генерации уникального номера заказа
std::string Order::generateOrderNumber() {
    srand(static_cast<unsigned int>(time(0)));// инициализация генератора случайных чисел
    int randomNum = rand() % 10000;// генерация случайного числа от 0 до 9999
    return "ORD" + std::to_string(randomNum);// создание номера
}

// Метод получения текущей даты и времени
std::string Order::getCurrentDate() {
    time_t now = time(0);// получение текущего времени
    // получение локального времени
#ifdef _MSC_VER
    struct tm timeinfo;
    localtime_s(&timeinfo, &now);// безопасное получение локального времени
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y %H:%M:%S", &timeinfo);// форматирование даты
#else
    tm* localTime = localtime(&now);// получение локального времени
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y %H:%M:%S", localTime);// форматирование даты
#endif
    return std::string(buffer);// возврат даты в виде строки
}

File: курсовая/c#/Order.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Text;
using System.Threading.Tasks;

namespace Курсовая_работа_Музыкальный_магазин_с_
{
    public class Order
    {
        private const int MAX_ORDER_ITEMS = 100;
        private string numberOrder;
        private string clientOrder;
        private Product[] orderItems;
        private int[] quantities;
        private int itemCount;
        private int countOrder;
        private double costOrder;
        private string statusOrder;
        private string payWay;
        private string shopOrder;
        private string dateOrder;

        // Конструкторы
        public Order()
        {
            orderItems = new Product[MAX_ORDER_ITEMS];
            quantities = new int[MAX_ORDER_ITEMS];
            itemCount = 0;
            countOrder = 0;
            costOrder = 0.0;

            numberOrder = GenerateOrderNumber();
            statusOrder = "Ожидает оплаты";
            dateOrder = GetCurrentDate();
            shopOrder = "Музыкальный магазин 'MusicLoverzz'";
            payWay = "Не выбран";
        }

        public Order(string clientId, Basket basket) : this()
        {
            clientOrder = clientId;
            CopyProductsFromBasket(basket);
        }

        // Деструктор
        ~Order()
        {
            ClearOrderItems();
        }

        public void ChangeStatus(string newStatus)
        {
            statusOrder = newStatus;
            if (newStatus == "Оплачен")
            {
                Console.WriteLine("\n=== ВАШ ЗАКАЗ УСПЕШНО ОФОРМЛЕН! ===");
                Console.WriteLine("Номер заказа: " + numberOrder);
                Console.WriteLine("Статус: " + statusOrder);
                Console.WriteLine("==================================");
            }
            else
            {
                Console.WriteLine("Статус заказа изменен на: " + statusOrder);
            }
        }

        public bool Payment()
        {
            Console.WriteLine("\n=== ОПЛАТА ЗАКАЗА ===");
            Console.WriteLine("Сумма к оплате: " + costOrder.ToString("F2") + " руб.");

            int paymentChoice = 0; // явная инициализация
            bool validChoice = false;

            while (!validChoice)
            {
                Console.WriteLine("Выберите способ оплаты:");
                Console.WriteLine("1. Банковская карта");
                Console.WriteLine("2. Электронный кошелек");
                Console.WriteLine("3. Наложенный платеж");
                Console.Write("Ваш выбор: ");

                string input = Console.ReadLine();

                if (int.TryParse(input, out paymentChoice))
                {
                    if (paymentChoice >= 1 && paymentChoice <= 3)
                    {
                        validChoice = true;
                    }
                    else
                    {
                        Console.WriteLine("Неверный выбор! Введите 1, 2 или 3.");
                    }
                }
                else
                {
                    Console.WriteLine("Ошибка! Введите число от 1 до 3.");
                }
            }


            if (paymentChoice == 1)
            {
                payWay = "Банковская карта";
            }
            else if (paymentChoice == 2)
            {
                payWay = "Электронный кошелек";
            }
            else 
            {
                payWay = "Наложенный платеж";
            }

            Console.WriteLine("Обработка платежа...");
            Console.WriteLine("Способ оплаты: " + payWay);
            Console.WriteLine("Оплата прошла успешно!");
            return true;
        }

        public void InfoOrder()
        {
            Console.WriteLine("\n=== ИНФОРМАЦИЯ О ЗАКАЗЕ ===");
            Console.WriteLine("Номер заказа: " + numberOrder);
            Console.WriteLine("Клиент: " + clientOrder);
            Console.WriteLine("Дата заказа: " + dateOrder);
            Console.WriteLine("Статус: " + statusOrder);
            Console.WriteLine("Способ оплаты: " + payWay);
            Console.WriteLine("Количество товаров: " + countOrder);
            Console.WriteLine("Общая стоимость: " + costOrder.ToString("F2") + " руб.");
            Console.WriteLine("Магазин: " + shopOrder);

            if (itemCount > 0)
            {
                Console.WriteLine("\nСостав заказа:");
                Console.WriteLine("------------------------------------");
                for (int i = 0; i < itemCount; i++)
                {
                    if (orderItems[i] != null)
                    {
                        Console.WriteLine($"{i + 1}. {orderItems[i].GetName()}" +
                                        $" - {quantities[i]} шт. x " +
                                        $"{orderItems[i].GetPrice():F2} руб. = " +
                                        $"{(quantities[i] * orderItems[i].GetPrice()):F2} руб.");
                    }
                }
                Console.WriteLine("------------------------------------");
            }
            Console.WriteLine("==================================");
        }

        public static bool MakeOrder(string clientId, Basket basket)
        {
            if (basket.IsEmpty())
            {
                Console.WriteLine("Корзина пуста! Добавьте товары перед оформлением заказа.");
                return false;
            }

            Console.WriteLine("\n=== ПОДТВЕРЖДЕНИЕ ЗАКАЗА ===");
            basket.DisplayBasket();
            Console.Write("\nВы точно хотите оформить заказ? (y/n): ");
            char confirmation = Console.ReadKey().KeyChar;
            Console.WriteLine();

            if (confirmation != 'y' && confirmation != 'Y')
            {
                Console.WriteLine("Оформление заказа отменено.");
                return false;
            }

            Order newOrder = new Order(clientId, basket);

            if (newOrder.Payment())
            {
                newOrder.ChangeStatus("Оплачен");
                Console.WriteLine("\n=== ЗАКАЗ УСПЕШНО ОФОРМЛЕН ===");
                newOrder.InfoOrder();
                newOrder.SaveToFile();
                basket.ClearBasket();
                return true;
            }
            else
            {
                newOrder.ChangeStatus("Ошибка оплаты");
                Console.WriteLine("Ошибка при оформлении заказа!");
                return false;
            }
        }

        public void SaveToFile()
        {
            using (StreamWriter file = new StreamWriter("orders.txt", true))
            {
                file.Write(numberOrder + "|" + clientOrder + "|" + statusOrder + "|" + costOrder.ToString("F2"));
                for (int i = 0; i < itemCount; i++)
                {
                    if (orderItems[i] != null)
                    {
                        file.Write("|" + orderItems[i].GetArticle() + "|" + quantities[i]);
                    }
                }
                file.WriteLine();
            }
        }

        // Геттеры
        public string GetNumberOrder() => numberOrder;
        public string GetClientOrder() => clientOrder;
        public string GetStatusOrder() => statusOrder;
        public double GetCostOrder() => costOrder;
        public int GetCountOrder() => countOrder;
        public string GetDateOrder() => dateOrder;
        public string GetPayWay() => payWay;
        public string GetShopOrder() => shopOrder;
        public int GetItemCount() => itemCount;

        public Product GetOrderItem(int index)
        {
            if (index >= 0 && index < itemCount)
            {
                return orderItems[index];
            }
            return null;
        }

        public int GetQuantity(int index)
        {
            if (index >= 0 && index < itemCount)
            {
                return quantities[index];
            }
            return 0;
        }

        private void CopyProductsFromBasket(Basket basket)
        {
            if (basket.IsEmpty())
            {
                return;
            }

            itemCount = basket.GetProductCount();
            countOrder = (int)basket.GetBasketCount();
            costOrder = basket.GetBasketCost();

            for (int i = 0; i < itemCount; i++)
            {
                Product basketProduct = basket.GetProduct(i);
                if (basketProduct != null)
                {
                    // Создаем новый Product вручную, копируя все поля
                    orderItems[i] = new Product(
                        basketProduct.GetArticle(),
                        basketProduct.GetName(),
                        basketProduct.GetBrand(),
                        basketProduct.GetProducerCountry(),
                        basketProduct.GetCategory(),
                        basketProduct.GetTypeProduct(),
                        basketProduct.GetDescription(),
                        basketProduct.GetQuantityStock(),
                        basketProduct.GetPrice()
                    );
                    quantities[i] = basket.GetQuantity(i);
                }
            }
        }

        private string GenerateOrderNumber()
        {
            Random rand = new Random();
            int randomNum = rand.Next(10000);
            return "ORD" + randomNum.ToString("D4");
        }

        private string GetCurrentDate()
        {
            return DateTime.Now.ToString("dd.MM.yyyy HH:mm:ss");
        }

        private void ClearOrderItems()
        {
            for (int i = 0; i < itemCount; i++)
            {
                orderItems[i] = null;
            }
            itemCount = 0;
        }
    }
}